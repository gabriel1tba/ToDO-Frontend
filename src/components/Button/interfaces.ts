export interface IButton extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
  loading?: boolean;
  color?: 'primary' | 'info' | 'warning' | 'danger' | 'success' | 'outline';
  icon?: JSX.Element;
  size?: 'large' | 'normal';
}
