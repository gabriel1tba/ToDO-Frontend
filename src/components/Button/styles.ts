import styled, { css } from 'styled-components';

interface WrapperButton {
  size: 'large' | 'normal';
  color: 'primary' | 'info' | 'warning' | 'danger' | 'success' | 'outline';
}

const colorModifiers = {
  primary: () => css`
    background: ${({ theme }) => theme.colors.primary.main};

    :hover {
      background: ${({ theme }) => theme.colors.primary.light};
    }
    :active {
      background: ${({ theme }) => theme.colors.primary.dark};
    }
    :disabled {
      background: ${({ theme }) => theme.colors.primary.dark};
      opacity: 0.6;
      cursor: not-allowed;
    }
  `,

  info: () => css`
    background: ${({ theme }) => theme.colors.info.main};

    :hover {
      background: ${({ theme }) => theme.colors.info.light};
    }
    :active {
      background: ${({ theme }) => theme.colors.info.dark};
    }
    :disabled {
      background: ${({ theme }) => theme.colors.info.dark};
      opacity: 0.6;
      cursor: not-allowed;
    }
  `,

  warning: () => css`
    background: ${({ theme }) => theme.colors.warning.main};

    :hover {
      background: ${({ theme }) => theme.colors.warning.light};
    }
    :active {
      background: ${({ theme }) => theme.colors.warning.dark};
    }
    :disabled {
      background: ${({ theme }) => theme.colors.warning.dark};
      opacity: 0.6;
      cursor: not-allowed;
    }
  `,

  success: () => css`
    background: ${({ theme }) => theme.colors.success.main};

    :hover {
      background: ${({ theme }) => theme.colors.success.light};
    }
    :active {
      background: ${({ theme }) => theme.colors.success.dark};
    }
    :disabled {
      background: ${({ theme }) => theme.colors.success.dark};
      opacity: 0.6;
      cursor: not-allowed;
    }
  `,

  danger: () => css`
    background: ${({ theme }) => theme.colors.danger.main};

    :hover {
      background: ${({ theme }) => theme.colors.danger.light};
    }
    :active {
      background: ${({ theme }) => theme.colors.danger.dark};
    }
    :disabled {
      background: ${({ theme }) => theme.colors.danger.dark};
      opacity: 0.6;
      cursor: not-allowed;
    }
  `,

  outline: () => css`
    background: transparent;
    border: 1px solid ${({ theme }) => theme.colors.gray[200]};
    color: ${({ theme }) => theme.colors.gray[500]};
  `,
};

const sizeModifiers = {
  normal: () => css`
    height: 38px;
    padding: 0.375rem 1rem;
    color: #fff;
    font-size: 1rem;
    line-height: 18px;
    font-weight: 400;
    border: 1px solid transparent;
    border-radius: 0.25rem;
  `,

  large: () => css`
    color: #fff;
    height: 52px;
    border-radius: 10px;
    border: 0;
    padding: 0 16px;
    width: 100%;
    font-weight: 500;

    font-size: 18px;
  `,
};

export const Wrapper = styled.button<WrapperButton>`
  display: flex;
  justify-content: center;
  align-items: center;

  transition: background 0.2s ease-in;

  svg {
    margin-left: 0.25rem;
  }

  ${({ size }) => sizeModifiers[size]()};

  ${({ color }) => colorModifiers[color]()};
`;

export const WrapperLoading = styled.div<Omit<WrapperButton, 'color'>>`
  display: flex;
  align-items: center;
  justify-content: center;

  > div {
    margin-left: ${({ size }) => (size === 'large' ? '24px' : '12px')};
  }
`;
