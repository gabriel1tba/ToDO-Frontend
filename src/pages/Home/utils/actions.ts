import { TTodo } from 'services/TodoService/interfaces';

export enum Types {
  GET_TODOS = 'GET_TODOS',
  CREATE_TODO = 'CREATE_TODO',
  UPDATE_TODO = 'UPDATE_TODO',
  DELETE_TODO = 'DELETE_TODO',
}

interface GetTodos {
  type: Types.GET_TODOS;
  payload: TTodo[];
}

interface CreateTodo {
  type: Types.CREATE_TODO;
  payload: TTodo;
}

interface UpdateTodo {
  type: Types.UPDATE_TODO;
  payload: TTodo;
}

interface DeleteTodo {
  type: Types.DELETE_TODO;
  payload: string;
}

export type Action = GetTodos | CreateTodo | UpdateTodo | DeleteTodo;

export const getTodosAction = (todos: TTodo[]): GetTodos => ({
  type: Types.GET_TODOS,
  payload: todos,
});

export const createTodoAction = (todo: TTodo): CreateTodo => ({
  type: Types.CREATE_TODO,
  payload: todo,
});

export const updateTodoAction = (todo: TTodo): UpdateTodo => ({
  type: Types.UPDATE_TODO,
  payload: todo,
});

export const deleteTodoAction = (id: string): DeleteTodo => ({
  type: Types.DELETE_TODO,
  payload: id,
});
