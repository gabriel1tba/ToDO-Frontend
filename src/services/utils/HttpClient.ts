import axios from 'axios';

const HttpClient = axios.create({
  baseURL: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`,
});

export default HttpClient;
