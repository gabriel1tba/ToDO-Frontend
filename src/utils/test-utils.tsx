import { render, RenderOptions } from '@testing-library/react';
import { TodoProvider } from 'context/todos';

interface IWrapperProvider {
  children: React.ReactNode;
}

const WrapperProvider = ({ children }: IWrapperProvider) => (
  <TodoProvider>{children}</TodoProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(<WrapperProvider>{ui}</WrapperProvider>, { ...options });

export * from '@testing-library/react';
export { customRender as render };
