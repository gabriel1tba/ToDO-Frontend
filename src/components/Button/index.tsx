import * as S from './styles';

import Loader from 'components/Loader';

import { IButton } from './interfaces';

const Button = ({
  children,
  loading,
  color = 'primary',
  size = 'normal',
  type = 'button',
  disabled,
  icon,
  ...rest
}: IButton) => (
  <S.Wrapper
    type={type}
    size={size}
    color={color}
    disabled={loading || disabled}
    {...rest}
  >
    {loading ? (
      <S.WrapperLoading size={size}>
        {children} <Loader size={12} color="white" />
      </S.WrapperLoading>
    ) : (
      <>
        {children} {icon}
      </>
    )}
  </S.Wrapper>
);

export default Button;
