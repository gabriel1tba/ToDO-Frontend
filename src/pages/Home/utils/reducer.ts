import { TTodo } from 'services/TodoService/interfaces';

import { Types, Action } from './actions';

const todosReducer = (state: TTodo[], action: Action): TTodo[] => {
  switch (action.type) {
    case Types.GET_TODOS: {
      return action.payload;
    }

    case Types.CREATE_TODO: {
      return [...state, action.payload];
    }

    case Types.UPDATE_TODO: {
      return state.map((todo) => {
        if (action.payload.id === todo.id) {
          todo = { ...action.payload, created_at: todo.created_at };
        }
        return todo;
      });
    }

    case Types.DELETE_TODO: {
      return state.filter((todo) => {
        return todo.id !== action.payload;
      });
    }

    default:
      return state;
  }
};

export default todosReducer;
