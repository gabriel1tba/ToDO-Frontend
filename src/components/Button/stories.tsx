import { Story, Meta } from '@storybook/react';
import { FaUser } from 'react-icons/fa';

import Button from '.';
import { IButton } from './interfaces';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string',
    },
    size: {
      type: 'string',
    },
    color: {
      type: 'string',
    },
    loading: {
      type: 'boolean',
    },
    icon: {
      type: 'function',
    },
  },
} as Meta;

export const Default: Story<IButton> = (args) => <Button {...args} />;

Default.args = {
  children: 'Without icon',
};

export const Large: Story<IButton> = (args) => <Button {...args} />;

Large.args = {
  children: 'Large',
  size: 'large',
};

export const Success: Story<IButton> = (args) => <Button {...args} />;

Success.args = {
  children: 'Success',
  color: 'success',
};

export const Loading: Story<IButton> = (args) => <Button {...args} />;

Loading.args = {
  children: 'Loading',
  loading: true,
};

export const Icon: Story<IButton> = (args) => <Button {...args} />;

Icon.args = {
  children: 'Icon',
  icon: <FaUser />,
};
