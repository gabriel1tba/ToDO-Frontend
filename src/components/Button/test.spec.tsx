import { render, screen } from 'utils/test-utils';
import { FiPlus } from 'react-icons/fi';

import Button from '.';

describe('<Button />', () => {
  it('should render the Button correctly without icon', () => {
    const { container } = render(<Button background="#007bff">Salvar</Button>);

    expect(screen.getByRole('button', { name: /salvar/i })).toBeInTheDocument();

    expect(container.firstChild).toHaveStyle({
      backgroundColor: '#007bff',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the Button correctly with icon icon', () => {
    render(
      <Button
        background="#007bff"
        icon={<FiPlus data-testid="add-icon" size={18} />}
      >
        Com ícone
      </Button>,
    );

    expect(
      screen.getByRole('button', { name: /com ícone/i }),
    ).toBeInTheDocument();

    expect(screen.getByTestId('add-icon')).toBeInTheDocument();
  });

  it('should render the Button correctly with icon icon', () => {
    render(
      <Button background="#007bff" loading>
        Com ícone
      </Button>,
    );

    expect(screen.getByText(/carregando.../i)).toBeInTheDocument();
  });

  it('should render the Button correctly with large size', () => {
    const { container } = render(
      <Button background="#007bff" size="large">
        Com ícone
      </Button>,
    );

    expect(container.firstChild).toHaveStyle({
      height: '52px',
      borderRadius: '10px',
      padding: '0 16px',
      width: '100%',
    });
  });
});
