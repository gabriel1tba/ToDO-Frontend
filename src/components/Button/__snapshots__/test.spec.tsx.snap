// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the Button correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: background 0.2s ease-in;
  transition: background 0.2s ease-in;
  height: 38px;
  padding: 0.375rem 1rem;
  color: #fff;
  font-size: 1rem;
  line-height: 18px;
  font-weight: 400;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  background: #ff9000;
}

.c0 svg {
  margin-left: 0.25rem;
}

.c0:hover {
  background: #FFA726;
}

.c0:active {
  background: #FF6D00;
}

.c0:disabled {
  background: #FF6D00;
  opacity: 0.6;
  cursor: not-allowed;
}

<button
  class="c0"
  type="button"
>
  Salvar
   
</button>
`;
